src/puppetlabs/puppetdb/catalogs.clj

```diff
@@ -114,7 +115,8 @@
    :version s/Str
    :environment (s/maybe s/Str)
    :transaction_uuid (s/maybe s/Str)
-   :producer_timestamp (s/maybe pls/Timestamp)
+   :producer_timestamp (s/pred #(not (nil? (to-timestamp %))) 'coercable-to-timestamp)
```

Comment:
We have at least three different internal representations of them in this
codebase: strings, java.sql.Date, and joda time. Whatever we do here should
(1) document what kind of data we expect and (2) kick out an error when we
get data that we can't deal with. I think this code gives that in the most
concrete possible way, by attempting the very conversion whose result we
care about.

That said: I assumed that pls/Timestamp was something stock, but it's
actually our own thing that does almost the exactly what I wrote down here.
(except with joda time, which is fine, since that's what clj-time converts
to internally in to-timestamp) So I suppose I'm ok just using that :)


On Wed, Mar 25, 2015 at 4:59 PM, Wyatt Alt <notifications@github.com> wrote:

> In src/puppetlabs/puppetdb/catalogs.clj
> <https://github.com/puppetlabs/puppetdb/pull/1311#discussion_r27178554>:
>
> > @@ -114,7 +115,8 @@
> >     :version s/Str
> >     :environment (s/maybe s/Str)
> >     :transaction_uuid (s/maybe s/Str)
> > -   :producer_timestamp (s/maybe pls/Timestamp)
> > +   :producer_timestamp (s/pred #(not (nil? (to-timestamp %))) 'coercable-to-timestamp)
>
> Actually, it looks like pls/Timestamp will also validate an integer so I
> retract the last bit there. I still feel pretty strongly about keeping
> these things as simple as possible though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetdb/pull/1311/files#r27178554>.
>